src/gui/dialogs/migrate_version_selection.cpp

```diff
@@ -0,0 +1,123 @@
+/*
+   Copyright (C) 2008 - 2018 by Jörg Hinrichs <joerg.hinrichs@alice-dsl.de>
+   Part of the Battle for Wesnoth Project https://www.wesnoth.org/
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY.
+
+   See the COPYING file for more details.
+*/
+
+#define GETTEXT_DOMAIN "wesnoth-lib"
+
+#include "gui/dialogs/migrate_version_selection.hpp"
+
+#include "addon/manager_ui.hpp"
+#include "filesystem.hpp"
+#include "game_version.hpp"
+#include "gui/auxiliary/find_widget.hpp"
+#include "gui/dialogs/migrate_version_selection.hpp"
+#include "gui/widgets/listbox.hpp"
+#include "gui/widgets/window.hpp"
+
+#include <boost/algorithm/string.hpp>
+
+static lg::log_domain log_version_migration{"gui/dialogs/migrate_version_selection"};
+#define ERR_LOG_VERSION_MIGRATION   LOG_STREAM(err,   log_version_migration)
+#define WRN_LOG_VERSION_MIGRATION   LOG_STREAM(warn,  log_version_migration)
+#define LOG_LOG_VERSION_MIGRATION   LOG_STREAM(info,  log_version_migration)
+#define DBG_LOG_VERSION_MIGRATION   LOG_STREAM(debug, log_version_migration)
+
+namespace gui2::dialogs
+{
+REGISTER_DIALOG(migrate_version_selection)
+
+void migrate_version_selection::execute()
+{
+    migrate_version_selection mig = migrate_version_selection();
+    if(mig.versions_.size() > 0)
+    {
+        mig.show();
+    }
+}
+
+migrate_version_selection::migrate_version_selection()
+{
+    version_info current_version = game_config::wesnoth_version;
+    std::string current_version_str = std::to_string(current_version.major_version())+"."+std::to_string(current_version.minor_version());
```

Comment:
These lines get repeated several times in this file, and they duplicate code that's already in filesystem.cpp's (currently private) `get_version_path_suffix` functions.

Reply:
Huh, alright. I did look for something already existing like that, but only in `version_info` rather than in `filesystem`.

Reply:
As a more general question, since I'm much more familiar with `static` in Java than C++: is there a reason that `filesystem.cpp` has essentially a bunch of private static functions in it?  I've also been running into various compiler/linker errors related to this, and examples I find online seem to all be based on static methods of a class, whereas here there is no "filesystem" class that these are attached to.

Reply:
When used outside a class or function, `static` means something is local to the current file – you can't reference it from another file.

Reply:
This use of `static` is the C-style equivalent of an putting the function in an anonymous namespace; I probably should have used the `namespace` keyword. When I added that code, I didn't consider that it might be used by the migration code (sorry).

Reply:
Well, the migration code wasn't going to exist for quite a long time yet anyway.  As long as there isn't a reason to keep it static though, I'll make it non-static and accessible to the migration dialog code.